packages/react-ui/src/features/pieces/lib/pieces-hook.ts

```diff
@@ -105,3 +67,73 @@ export const piecesHooks = {
     });
   },
 };
+
+function stepMetadataQueryBuilder({ step }: { step: Action | Trigger }) {
+  const isPieceStep =
+    step.type === ActionType.PIECE || step.type === TriggerType.PIECE;
+  const pieceName = isPieceStep ? step.settings.pieceName : `${step.type}`;
+  const pieceVersion = isPieceStep ? step.settings.pieceVersion : ``;
+  const type = step.type;
+  return {
+    queryKey: ['piece', type, pieceName, pieceVersion],
+    queryFn: async () => {
+      const metadata = await getStepMetadata(type, pieceName, pieceVersion);
+      return {
+        ...metadata,
+        type,
+        pieceName,
+        pieceVersion,
+      };
+    },
+    staleTime: Infinity,
+  };
+}
+
+async function getStepMetadata(
+  type: ActionType | TriggerType,
+  pieceName: string,
+  pieceVersion: string,
+): Promise<StepMetadata> {
+  const metadataExtractor = async () => {
```

Comment:
I don't think this should be function inside function, I don't understand how pieceName and pieceVersion is always defined for types that are not PIECE.